samples/300/connect_retry_with_telemetry.py

```diff
@@ -0,0 +1,171 @@
+CONNECTION_STRING = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
+
+
+# The interval at which to send telemetry
+TELEMETRY_INTERVAL = 5
+# Number of times connection attempts will be made.
+RETRY_NOS = 50
+# Interval between consecutive connection attempts
+ATTEMPT_INTERVAL = 5
+connected_event = None
+disconnected_event = None
+main_event_loop = None
+
+# TODO We might have to expand our error list to be more detailed
+# And possibly have status codes that reflect what sort of a service error
+# To have a more detailed experience
+transient_errors = [
+    exceptions.OperationCancelled,
+    exceptions.OperationTimeout,
+    exceptions.ServiceError,
+    exceptions.ConnectionFailedError,
+    exceptions.ConnectionDroppedError,
+    exceptions.NoConnectionError,
+    exceptions.ClientError,  # TODO Only TLSEsxhangeError is actually worth retrying
+]
+
+
+def is_retryable(exc):
+    if type(exc) in transient_errors:
+        return True
+    else:
+        return False
+
+
+def create_client():
+    global connected_event
+    global disconnected_event
+
+    # Create a Device Client
+    device_client = IoTHubDeviceClient.create_from_connection_string(CONNECTION_STRING)
+
+    connected_event = asyncio.Event()
+    disconnected_event = asyncio.Event()
+
+    # Define a connection state change handler
+    async def handle_on_connection_state_change():
+        global main_event_loop
+        print(
+            "handle_on_connection_state_change fired. Connected status : {}".format(
+                device_client.connected
+            )
+        )
+        if device_client.connected:
+            print("Connected event is set...")
+            main_event_loop.call_soon_threadsafe(connected_event.set)
+        else:
+            print("Disconnected event is set...")
+            main_event_loop.call_soon_threadsafe(disconnected_event.set)
+
+    try:
+        # Attach the connection state handler
+        device_client.on_connection_state_change = handle_on_connection_state_change
+    except Exception:
+        # Clean up in the event of failure
+        device_client.shutdown()
+        raise
+
+    return device_client
+
+
+async def do_work(device_client):
+    global connected_event
+    print("Client is connected {}".format(device_client.connected))
+    if not device_client.connected:
+        print("Waiting for connection ...")
+        await connected_event.wait()
+
+    item = random.Random()
+    print("Sending message...")
+    try:
+        await device_client.send_message(str(item))
+        print("Message sent...")
+        await asyncio.sleep(TELEMETRY_INTERVAL)
+    except Exception:
+        print("Caught exception while trying to send message....")
+        logging.debug(traceback.format_exc())
+
+
+async def connect_with_retry(device_client, number_of_tries=None):
+    if device_client.connected:
+        print("Client is already connected...")
```

Comment:
Because connection retry is enabled in this sample (by default), it seems unnecessary to check the connection and issue a manual connect every pass of the loop in main. This function seems like it should be invoked outside of that loop just for the initial connect.

Reply:
i had once had the initial connect outside but i had to still catch exception and retry even at that time - so does not make sense to put it outside. I just bypass the connection logic if it is already connected.